### PR TITLE
Pull in build revamp for prebuilt targets

### PR DESCRIPTION
**Pull request by computerquip-streamlabs**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,85 +1,47 @@
...
-	${CMAKE_JS_LIB}
-)
+	${NODEJS_LIBRARIES})
+
+target_include_directories(node_libuiohook
+	PUBLIC ${NODEJS_INCLUDE_DIRS} ${UIOHOOKPATH}/include/)
+
+install(
+	FILES $<TARGET_PDB_FILE:node_libuiohook>
+	DESTINATION ${CMAKE_INSTALL_PREFIX} OPTIONAL)
```
Comment (Xaymar) [inline CMakeLists.txt]: ${CMAKE_INSTALL_PREFIX} should be "./" (or just "/").

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,85 +1,47 @@
...
 
-add_library(node-libuiohook_THIRDPARTY SHARED IMPORTED)
+install(
+	FILES $<TARGET_FILE:node_libuiohook>
+	DESTINATION ${CMAKE_INSTALL_PREFIX})
```
Comment (Xaymar) [inline CMakeLists.txt]: ${CMAKE_INSTALL_PREFIX} should be "./" (or just "/").

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,85 +1,47 @@
...
-	"${node-libuiohook_BINARY_DIR}/distribute/node-libuiohook/$<TARGET_FILE_NAME:node-libuiohook>"
-)
-ADD_CUSTOM_COMMAND(TARGET node-libuiohook POST_BUILD
-	COMMAND ${CMAKE_COMMAND} -E copy
-	${node-libuiohook_THIRDPARTY}
-	"${node-libuiohook_BINARY_DIR}/distribute/node-libuiohook"
-)
+install(
+	FILES ${node_libuiohook_EXTRAS}
+	DESTINATION ${CMAKE_INSTALL_PREFIX})
```
Comment (Xaymar) [inline CMakeLists.txt]: ${CMAKE_INSTALL_PREFIX} should be "./" (or just "/").

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,85 +1,47 @@
-cmake_minimum_required(VERSION 3.5)
-project(node-libuiohook)
+cmake_minimum_required(VERSION 3.0)
+project(node_libuiohook)
```
Comment (Xaymar) [inline CMakeLists.txt]: What is the reason behind changing the project name to have an underscore instead of a dash?

**Review by Xaymar — CHANGES_REQUESTED** [review]
Minor changes should be applied to the install commands.

File: CMakeLists.txt (hunk 1 above)
Comment (Xaymar) [reply CMakeLists.txt]: See documentation for DESTINATION: https://cmake.org/cmake/help/latest/command/install.html?highlight=DESTINATION

**Comment by computerquip-streamlabs** [comment]
Node modules aren't allowed to have dashes in their name. 
In addition, CMAKE_INSTALL_PREFIX isn't a relative path, it's the full path to the directory I want. If I specify a relative path, it will result in the exact same thing with a weird path with either an extra slash or dot that does nothing. They do the essentially the same thing but one gives a cleaner output.

Why do you think you're required to specify a relative path here?

**Comment by Xaymar** [comment]
> Why do you think you're required to specify a relative path here?

Quoting the documentation: 

> If a relative path is given it is interpreted relative to the value of the CMAKE_INSTALL_PREFIX variable. The prefix can be relocated at install time using the DESTDIR mechanism explained in the CMAKE_INSTALL_PREFIX variable documentation.

Not using a relative path here actually breaks other parts that rely on the INSTALL command to be relative, such as CPack. I have encountered this issue before already in enc-amf, enc-aomedia-av1 and obs-stream-effects.

> Node modules aren't allowed to have dashes in their name.

The documentation for node does not state this, the only documentation that I found that restricts package names was from npm (https://docs.npmjs.com/files/package.json), which also does not state that dashes are disallowed. The only known limitation would be RFC 3986, which defines what is a unreserved character for URIs, see [here](http://www.ietf.org/rfc/rfc3986.txt). And a dash/hyphen is not condered a reserved character, thus is allowed in the name.

However what the npm documentation does say is that you shouldn't include node in the package name, as it is pretty clear that your module is going to be for node or similar.

Additionally we are already using modules with a dash in their name or even two.

**Comment by computerquip-streamlabs** [comment]
uh, just realized that the module name is actually broken here as well. Node requires the name specified to `NODE_MODULE` be the same as the actual filename. It states: `The module_name must match the filename of the final binary (excluding the .node suffix).` EDIT: In case it's not obvious here, `NODE_MODULE` requires the name be a valid C identifier which can't have dashes. It also didn't (doesn't?) accept quotes as it stringifies the identifier.

That said, we have been using it with dashes and different names previously to seemingly no negative effect. I'm going to post an issue on node's repository and see what type of response I get. 

As for the CPack comment, sure. Didn't realize CPack expected a relative path.

**Comment by computerquip-streamlabs** [comment]
I'll see what the result of this is: https://github.com/nodejs/node/issues/21961

**Comment by Xaymar** [comment]
```
NODE_MODULE(obs_studio_node, main);
```

would resolve to

```
NODE_MODULE_X(obs_studio_node, main, NULL, 0);
```

would resolve to

```
extern "C" {
    static node::node_module _module =
    {
      NODE_MODULE_VERSION,
      flags,
      NULL,
      __FILE__,
      (node::addon_register_func) (main),
      NULL,
      NODE_STRINGIFY(obs_studio_node),
      priv,
      NULL
    };
    NODE_C_CTOR(_register_ ## obs_studio_node) {
      node_module_register(&_module);
    }
  }
```

The problem here is NODE_C_CTOR, which defines an exported auto-init function. This is actually what is exported, and why you can't have a - in there. I'm not sure why they didn't stick with a _register_node function and instead have things named if it is actually a limitation.

**Comment by Xaymar** [comment]
Found the code that does the actual loading on node.js v10:

```
using InitializerCallback = void (*)(Local<Object> exports,
                                     Local<Value> module,
                                     Local<Context> context);

inline InitializerCallback GetInitializerCallback(DLib* dlib) {
  const char* name = "node_register_module_v" STRINGIFY(NODE_MODULE_VERSION);
  return reinterpret_cast<InitializerCallback>(dlib->GetSymbolAddress(name));
}

inline napi_addon_register_func GetNapiInitializerCallback(DLib* dlib) {
  const char* name =
      STRINGIFY(NAPI_MODULE_INITIALIZER_BASE) STRINGIFY(NAPI_MODULE_VERSION);
  return
      reinterpret_cast<napi_addon_register_func>(dlib->GetSymbolAddress(name));
}

static void DLOpen(const FunctionCallbackInfo<Value>& args) {
//...
  // -1 is used for N-API modules
  if ((mp->nm_version != -1) && (mp->nm_version != NODE_MODULE_VERSION)) {
    // Even if the module did self-register, it may have done so with the wrong
    // version. We must only give up after having checked to see if it has an
    // appropriate initializer callback.
    if (auto callback = GetInitializerCallback(&dlib)) {
      callback(exports, module, context);
      return;
    }
    char errmsg[1024];
    snprintf(errmsg,
             sizeof(errmsg),
             "The module '%s'"
             "\nwas compiled against a different Node.js version using"
             "\nNODE_MODULE_VERSION %d. This version of Node.js requires"
             "\nNODE_MODULE_VERSION %d. Please try re-compiling or "
             "re-installing\nthe module (for instance, using `npm rebuild` "
             "or `npm install`).",
             *filename, mp->nm_version, NODE_MODULE_VERSION);

    // NOTE: `mp` is allocated inside of the shared library's memory, calling
    // `dlclose` will deallocate it
    dlib.Close();
    env->ThrowError(errmsg);
    return;
  }
  if (mp->nm_flags & NM_F_BUILTIN) {
    dlib.Close();
    env->ThrowError("Built-in module self-registered.");
    return;
  }

  mp->nm_dso_handle = dlib.handle_;
  mp->nm_link = modlist_addon;
  modlist_addon = mp;

  if (mp == nullptr) {
    if (auto callback = GetInitializerCallback(&dlib)) {
      callback(exports, module, context);
    } else if (auto napi_callback = GetNapiInitializerCallback(&dlib)) {
      napi_module_register_by_symbol(exports, module, context, napi_callback);
    } else {
      dlib.Close();
      env->ThrowError("Module did not self-register.");
    }
    return;
  }
//...
}
```

So in v10 node.js, the function is either N-API or just called node_register_module_vVERSIONHERE.

**Comment by Xaymar** [comment]
Same function on v8:

```
static void DLOpen(const FunctionCallbackInfo<Value>& args) {
  Environment* env = Environment::GetCurrent(args);

  CHECK_EQ(modpending, nullptr);

  if (args.Length() < 2) {
    env->ThrowError("process.dlopen needs at least 2 arguments.");
    return;
  }

  int32_t flags = DLib::kDefaultFlags;
  if (args.Length() > 2 && !args[2]->Int32Value(env->context()).To(&flags)) {
    return env->ThrowTypeError("flag argument must be an integer.");
  }

  Local<Object> module =
      args[0]->ToObject(env->context()).ToLocalChecked();  // Cast
  node::Utf8Value filename(env->isolate(), args[1]);  // Cast
  DLib dlib;
  dlib.filename_ = *filename;
  dlib.flags_ = flags;
  bool is_opened = dlib.Open();

  // Objects containing v14 or later modules will have registered themselves
  // on the pending list.  Activate all of them now.  At present, only one
  // module per object is supported.
  node_module* const mp = modpending;
  modpending = nullptr;

  if (!is_opened) {
    Local<String> errmsg = OneByteString(env->isolate(), dlib.errmsg_.c_str());
    dlib.Close();
#ifdef _WIN32
    // Windows needs to add the filename into the error message
    errmsg = String::Concat(errmsg,
                            args[1]->ToString(env->context()).ToLocalChecked());
#endif  // _WIN32
    env->isolate()->ThrowException(Exception::Error(errmsg));
    return;
  }

  if (mp == nullptr) {
    dlib.Close();
    env->ThrowError("Module did not self-register.");
    return;
  }
  if (mp->nm_version == -1) {
    if (env->EmitNapiWarning()) {
      ProcessEmitWarning(env, "N-API is an experimental feature and could "
                         "change at any time.");
    }
  } else if (mp->nm_version != NODE_MODULE_VERSION) {
    char errmsg[1024];
    snprintf(errmsg,
             sizeof(errmsg),
             "The module '%s'"
             "\nwas compiled against a different Node.js version using"
             "\nNODE_MODULE_VERSION %d. This version of Node.js requires"
             "\nNODE_MODULE_VERSION %d. Please try re-compiling or "
             "re-installing\nthe module (for instance, using `npm rebuild` "
             "or `npm install`).",
             *filename, mp->nm_version, NODE_MODULE_VERSION);

    // NOTE: `mp` is allocated inside of the shared library's memory, calling
    // `dlclose` will deallocate it
    dlib.Close();
    env->ThrowError(errmsg);
    return;
  }
  if (mp->nm_flags & NM_F_BUILTIN) {
    dlib.Close();
    env->ThrowError("Built-in module self-registered.");
    return;
  }

  mp->nm_dso_handle = dlib.handle_;
  mp->nm_link = modlist_addon;
  modlist_addon = mp;

  Local<String> exports_string = env->exports_string();
  MaybeLocal<Value> maybe_exports =
      module->Get(env->context(), exports_string);

  if (maybe_exports.IsEmpty() ||
      maybe_exports.ToLocalChecked()->ToObject(env->context()).IsEmpty()) {
    dlib.Close();
    return;
  }

  Local<Object> exports =
      maybe_exports.ToLocalChecked()->ToObject(env->context())
          .FromMaybe(Local<Object>());

  if (mp->nm_context_register_func != nullptr) {
    mp->nm_context_register_func(exports, module, env->context(), mp->nm_priv);
  } else if (mp->nm_register_func != nullptr) {
    mp->nm_register_func(exports, module, mp->nm_priv);
  } else {
    dlib.Close();
    env->ThrowError("Module has no declared entry point.");
    return;
  }

  // Tell coverity that 'handle' should not be freed when we return.
  // coverity[leaked_storage]
}
```

I don't see _register being called, which means it's probably being called automatically on load through the CRT handler added through NODE_C_CTOR. Which in turn means that the actual name is mostly ignored, except for here:

```
inline struct node_module* FindModule(struct node_module* list,
                                      const char* name,
                                      int flag) {
  struct node_module* mp;

  for (mp = list; mp != nullptr; mp = mp->nm_link) {
    if (strcmp(mp->nm_modname, name) == 0)
      break;
  }

  CHECK(mp == nullptr || (mp->nm_flags & flag) != 0);
  return mp;
}
```

This is used to find already loaded modules, which means that `require('obs-studio-node')` won't match `NODE_MODULE(obs_studio_node)` from what I've understood.

**Comment by computerquip-streamlabs** [comment]
I'm starting to think this doesn't matter and we should pass whatever to MODULE_NAME and match the filename... the actual module name require'd from Javascript is something different anyways. I don't like the idea of breaking a requirement because it practically doesn't matter but we can avoid it either way regardless.

**Comment by Xaymar** [comment]
From what I found out, it doesn't actually matter except for the module name. If you replace the macro with your own in which you replace what the initializer function is named you can have your module called anything that is a valid file name on Linux, Mac and Windows.

Though I didn't find anything that directly tried to compare names other than that function, so maybe even the module name is ignored. v10 didn't seem to check it.

**Pull request merged**